src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -518,20 +522,42 @@ public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBas
             }
         }
 
+        private static readonly ObjectPool<CheckConstraintsArgsBoxed> s_checkConstraintsArgsBoxedPool = new ObjectPool<CheckConstraintsArgsBoxed>(() => new CheckConstraintsArgsBoxed());
+
         internal sealed class CheckConstraintsArgsBoxed
         {
-            public readonly CheckConstraintsArgs Args;
+            public CheckConstraintsArgs Args;
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
             {
-                Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                var boxedArgs = s_checkConstraintsArgsBoxedPool.Allocate();
+                boxedArgs.Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                return new PooledCheckConstraintsArgsBoxed(boxedArgs);
             }
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            {
+                var boxedArgs = s_checkConstraintsArgsBoxedPool.Allocate();
+                boxedArgs.Args = new CheckConstraintsArgs(currentCompilation, conversions, includeNullability, location, diagnostics);
+                return new PooledCheckConstraintsArgsBoxed(boxedArgs);
+            }
+
+            internal ref struct PooledCheckConstraintsArgsBoxed
             {
-                Args = new CheckConstraintsArgs(currentCompilation, conversions, includeNullability, location, diagnostics);
+                public readonly CheckConstraintsArgsBoxed BoxedArgs;
+
+                public PooledCheckConstraintsArgsBoxed(CheckConstraintsArgsBoxed checkConstraintsArgs)
+                {
+                    BoxedArgs = checkConstraintsArgs;
+                }
+
+                public void Dispose()
```

Comment:
Sorry, but the compiler team discussed the use of disposal for free'ing pooled objects and did not support it.
Unless I'm missing a difference, the same rationale still applies:
[https://github.com/dotnet/roslyn/pull/67198#discussion_r1128075922](https://github.com/dotnet/roslyn/pull/67198#discussion_r1128075922)

Reply:
this is not Dispose on the object itself.  It's Disposal on the *owner* (which is fine.  the owner just takes the object and puts back in the pool).

Reply:
(This can also be safe, as we can null out BoxedArgs if people are worried about double-frees or something like that).

Reply:
Unless there are exceptions involved, we don't need the `try`. A simple call to `Free()` seems adequate. I would rather not re-hash the arguments :-/

Reply:
The 'try' is needed as nothing stops early return (i just found a case of this in another PR i was doing, where we weren't returning pooled array builders in teh compielr because of htat).  

I'm not rehashing any arguments here.  I'm stating that hte existing position is not violated by this PR because the pooled object itself is not freeing itself.  This is a normal case of an owner type that is disposable which releases *other* resources safely** to a pool.

(technically it could be safer by nulling things out to prevent any double-dispose issues, but that's orthogonal to my point).

Reply:
Moved to prevent any chance of double-dispose.